enhancements/machine-config/onprem-namespaces.md

```diff
@@ -0,0 +1,134 @@
+---
+title: namespaces-for-on-prem-networking-services
+authors:
+  - "@cybertron"
+reviewers:
+  - "@smarterclayton"
+  - "@celebdor"
+  - "@mandre"
+  - "@rgolangh"
+  - "@patrickdillon"
+  - "@kikisdeliveryservice"
+approvers:
+  - Someone from MCO
+  - Someone from architecture
+creation-date: 2020-07-24
+last-updated: 2020-07-27
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Namespaces for On-Prem Networking Services
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+When support for on-premise deployments on baremetal, OpenStack, OVirt, and
+VSphere was added, some additional networking services were needed to provide
+functionality that would come from cloud services on other platforms. These
+services were added in namespaces called openshift-kni-infra,
+openshift-openstack-infra, etc. There are two problems with this: 1) The
+namespaces are created in all deployments, including on platforms other
+than the one that uses the namespace and 2) The openshift-x-infra name scheme
+does not match the existing pattern for OpenShift namespaces.
+
+## Motivation
+
+The existence of namespaces not related to a given deployment is confusing to
+users. Additionally, the inconsistent naming of the namespaces is a further
+possible point of confusion.
+
+### Goals
+
+- Only create namespaces for networking services on deployments that need them.
+- Use names that are consistent with the rest of OpenShift.
+
+### Non-Goals
+
+## Proposal
+
+Instead of creating the namespaces at install time as is currently done, add
+them to the list of
+[resources to be synced](https://github.com/openshift/machine-config-operator/blob/master/pkg/operator/sync.go)
+in the machine-config-operator. This will allow us to add logic that considers
+the platform before creating them.
+
+For the naming, we propose `openshift-hosted-network-services` for all
+platforms that use these components. There is no need for a platform-specific
+name since only one platform can exist at a time anyway.
```

Comment:
I don't know why but by reading this, my thought was: "should we have a dedicated operator that does just this and have the MCO help with, perhaps, deploying the needed files when it has to"? This is more to have the MCO being the API for system configuration and other components leverage it w/o adding anything super specific to it. The counter-argument would be "but you already ship per-platform configurations". Right, we do but we have specialized controllers (kube, crio) that handle the lifecycle of these configurations and the lifecycle of other configurations is simple enough that an upgrade and an overwrite is sufficient. For things like these network services it may warrant a better lifecycle handler to me since those are effectively namespaced services.

I'm committing to finding a very old conversation where I'm pretty sure shipping these services (and ns) was something which we would have iterated on to find them a definite better place.
I'm familiar with this namespace issue/bz and what we're trying to solve but I think a better design, if possible, would be towards finding a better home, specialized for all these. Could be something like what happened with etcd-quorum-guard, here, we're talking about network services for on-premise platforms which sounds like a great specialization that we can provide a better and dedicated tool to handle the task. Should we explore a path like this?

I guess what I'm trying to say/ask is should we re-focus on a broader goal w/o focusing on the namespaces only which already sound like something "external" or pretty much a "side" of the MCO? I'm happy to help scope the issue further and provide any MCO help needed (even on our side if this helps with having a better MCO API that others can use w/o plugging directly into the MCO)

Reply:
I guess I've gotten used to those things living in the MCO repo, but as they get refined a baremetal & friends (openshift-hosted-network-services) network configs operator works too and makes a lot of sense - great points @runcom 
 
IIRC the namespaces were eventually going to be moved to a better place? Could make more sense to make that move now rather than shoehorning this into the MCO as this seems to expand over time and overloading the MCO. Might be better suited in a dedicated operator that handles the namespace & networking for onprem.

Reply:
I think the idea of separate operator has been in the pipeline since the inception, but we know there are pretty big hurdles to get there,

- work up a workflow that works for bootstrapping no dependent on scheduling a workload on control-plane because we need these services to run for the node to join the cluster itself.. and today the only mechanism is MCO and machine config.
- If the operator wants to manage the lifecycle of these components using machine configs, that means that we need some way during upgrades to syncronize the update of other configuration of machine with the changes as required by new version of the operator.. this is tricky because of reboots to apply the change by machine configs.
- If the operator decides to manage the components itself by manipulating the files on disk like other static-pod-operators like etcd, kube-apiserver, kube-scheduler, kube-controller-manger {operator} that means that there is one workflow to install using machine configs but lifecycle is managed  by different process.

Figuring out and getting this separate operator is not going to be easy and quick. So as long as we know how difficult it would be to move to stand alone operator we need to weigh that to how quickly we need to get the namespaces fixed.

And we could probably move namespaces and then work on migrating to operator.

Reply:
Q is how much effort this implementation will be vs operator and if, after this implementation, any will to do the operator actually exists to make the bigger jump... things tend to expand and never leave the MCO.

Reply:
> Figuring out and getting this separate operator is not going to be easy and quick. So as long as we know how difficult it would be to move to stand alone operator we need to weigh that to how quickly we need to get the namespaces fixed.
> 
> And we could probably move namespaces and then work on migrating to operator.

as Kirsten already asked, I think it's crucial to understand a path to get there. I would love to hear more on the priority of moving those namespaces though, we've been living with them for quite a while now, I'd be surprised by the urgent priority all of sudden, which makes me think that we might better use our time to tackle the broader problem. I think it has been said that "moving the namespaces" isn't as easy as a rename and as Erica pointed out above, this requires logic in the MCO to support a migration, which again tells me we might pursue the broader issue.

Reply:
From our perspective, we're trying to respond to https://bugzilla.redhat.com/show_bug.cgi?id=1815561 appropriately based on the priority it was assigned and the timeframe presented in the description (which we've already missed). I tend to agree that it's a largely cosmetic issue, but it isn't up to me. :-)

I believe that as a team, we would still like to pursue the separate operator approach at some point. However, as Abhinav noted above that's going to be a large piece of work and it's not likely to happen in the near or even medium term, if at all given the response to the last new operator we proposed. I think recent discussions have raised questions for me as to whether it even makes sense to create a new operator that is going to be that tightly coupled to MCO.

I'm happy to have these discussions, but I'm pretty sure I don't have the authority to indefinitely postpone fixing this bug so if that's the way we choose to go I need someone else to make that call.